iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+
+  Ptr<ProxyMutex> lock; // Lock
+
+private:
+  void metric_inc(RefCountCache_Stats metric_enum, int64_t data);
+
+  unsigned int part_num;
+
+  uint64_t max_size;
+  unsigned int max_items;
+
+  uint64_t size;
+  unsigned int items;
+
+  hash_type item_map;
+
+  PriorityQueue<RefCountCacheHashEntry *> expiry_queue;
+  RecRawStatBlock *rsb;
+};
+
+template <class C>
+RefCountCachePartition<C>::RefCountCachePartition(unsigned int part_num, uint64_t max_size, unsigned int max_items,
+                                                  RecRawStatBlock *rsb)
+{
+  this->part_num  = part_num;
+  this->max_size  = max_size;
+  this->max_items = max_items;
+  this->size      = 0;
+  this->items     = 0;
+  this->rsb       = rsb;
+
+  // Initialize lock
+  this->lock = new_ProxyMutex();
+}
+
+template <class C>
+Ptr<C>
+RefCountCachePartition<C>::get(uint64_t key)
+{
+  this->metric_inc(refcountcache_total_lookups_stat, 1);
+  location_type l = this->item_map.find(key);
+  if (l.isValid()) {
+    // found
+    this->metric_inc(refcountcache_total_hits_stat, 1);
+    return make_ptr((C *)l.m_value->item.get());
+  } else {
+    return Ptr<C>();
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::put(uint64_t key, C *item, int size, int expire_time)
+{
+  this->metric_inc(refcountcache_total_inserts_stat, 1);
+  size += sizeof(C);
+  // Remove any colliding entries
+  this->erase(key);
+
+  // if we are full, and can't make space-- then don't store the item
+  if (this->is_full() && !this->reserve(size)) {
+    Debug("refcountcache", "partition %d is full-- not storing item key=%ld", this->part_num, key);
+    this->metric_inc(refcountcache_total_failed_inserts_stat, 1);
+    return;
+  }
+
+  // Create our value-- which has a ref to the `item`
+  RefCountCacheHashEntry *val = refCountCacheHashingValueAllocator.alloc();
+  val->set(item, key, size, expire_time);
+
+  // add expiry_entry to expiry queue, if the expire time is positive (otherwise it means don't expire)
+  if (expire_time >= 0) {
+    Debug("refcountcache", "partition %d adding entry with expire_time=%d\n", this->part_num, expire_time);
+    PriorityQueueEntry<RefCountCacheHashEntry *> *expiry_entry = expiryQueueEntry.alloc();
+    new ((void *)expiry_entry) PriorityQueueEntry<RefCountCacheHashEntry *>(val);
+    expiry_queue.push(expiry_entry);
+    val->expiry_entry = expiry_entry;
+  }
+
+  // add the item to the map
+  this->item_map.insert(val);
+  this->size += val->meta.size;
+  this->items++;
+  this->metric_inc(refcountcache_current_size_stat, (int64_t)val->meta.size);
+  this->metric_inc(refcountcache_current_items_stat, 1);
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::erase(uint64_t key, int expiry_time)
+{
+  location_type l = this->item_map.find(key);
+  if (l.isValid()) {
+    if (expiry_time >= 0 && l.m_value->meta.expiry_time != expiry_time) {
+      return;
+    }
+    // found
+    this->item_map.remove(l);
```

Comment:
This is somewhat confusing. ``TSHashMap`` sort of assumes that removing items also frees them (at least it does in the remove by free path). However since you are using ``DLL`` removed items need to be manually freed. Deserves some code comments I think.